Merge WooCommerce Brands into core

jimjasson opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR rolls the previously paid [WooCommerce Brands](https://woocommerce.com/products/brands/) plugin into WooCommerce core, ensuring that all WooCommerce Brands features are now available for free. The Brands functionality is enabled via the `Automattic\WooCommerce\Internal\Brands` class.

When Brands functionality is enabled both by core and by the WooCommerce Brands plugin, then core will deactivate the standalone extension and advise users to delete it. Moreover, a notice will show up for users who already have Brands installed but not activated. Finally, it won't be possible to enable Brands via WP CLI. 


The plan is to release Brands:
- to all users behind a feature flag in v9.4,
- to all users in v9.5 but have it enabled only for 5% and then;
- enable it for all users in v9.6.

#### Technical details 

Important: 

~~The approach we followed with this merge is to keep all Brands functionality in separate files (with minimal exceptions) to ensure a smooth transition from the Brands plugin to core. However, in the future, it is worth exploring the option of moving away from a hook-based system and instead move the Brands code inside existing files in core. For example, instead of having a separate file for Brands restrictions for coupons, we could merge this logic inside the existing coupons class.~~

The approach we followed with this merge is to keep all Brands functionality in separate files, so that the Brands module can be disabled by users who don't want to use it. 

Brands functionality is enabled via the `Automattic\WooCommerce\Internal\Brands` class. This class includes: 
- `includes/class-wc-brands.php`, which is the class responsible for registering the Brands taxonomy, initializing the Brands widgets and shortcodes. 
- `includes/class-wc-brands-coupons.php`, which is the class responsible for adding a Brands usage restriction to coupons. 
- `includes/class-wc-brands-brand-settings-manager.php`, which is a helper class designed to get/set Brands restrictions on coupons. In the future this class could be merged into `includes/class-wc-brands-coupons.php`.
- `includes/wc-brands-functions.php`, which contains Brands helper functions. 
- `includes/blocks/class-wc-brands-block-templates.php` and `includes/blocks/class-wc-brands-block-template-utils-duplicated.php` which are the classes responsible for rendering the Brand block when an FSE theme is used. 
- `includes/admin/class-wc-admin-brands.php`, which is the class that is responsible for displaying and saving all Brands admin settings. 

Brands also have an integration with v1 and v2 of the REST API. The controllers for Brands can be found under:
- `includes/rest-api/Controllers/Version1/class-wc-rest-product-brands-v1-controller.php`
- `includes/rest-api/Controllers/Version2/class-wc-rest-product-brands-v2-controller.php`.

As for styles:
- all frontend styles have been added in a new `woocommerce/client/legacy/css/brands.scss` stylesheet.
- all admin styles have been added to the new `woocommerce/client/legacy/css/brands-admin.scss` stylesheet. 

As for scripts:
- the only Brands script has been added in `woocommerce/client/legacy/js/admin/wc-brands-enhanced-select.js`, which gets enqueued from: `woocommerce/includes/admin/class-wc-admin-brands.php` along with the admin styles. 

As for the new Product Editor integration:
- the Brands/Product Editor compatibility is loaded via the `plugins/woocommerce/includes/class-wc-brands.php` class.

As for unit tests:
- the existing core's REST API unit tests have been updated to expect 'brands' as a valid REST API field and;
- a new `woocommerce/tests/php/includes/admin/class-wc-admin-brands-test.php` unit test has been added to test the Brands admin flows. 

As for E2E tests:
- a new `woocommerce/tests/e2e-pw/tests/merchant/create-product-brand.spec.js` E2E test has been added to cover the creation, editing and deletion of a product brand and; 
- the existing `woocommerce/tests/e2e-pw/tests/merchant/create-restricted-coupons.spec.js` test has been updated to include a test about Brands restrictions.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Part 1: Activation

1. Use WooCommerce `trunk` or the latest stable release.
2. Install WooCommerce Brands.
3. Following the WooCommerce Brands [documentation](https://woocommerce.com/document/woocommerce-brands/#creating-brands), go to the **Products > Brands** tab and create some Brands.
4. Go to the Plugins tab.
5. Check out this branch. 
6. Update `update_option( 'wc_feature_woocommerce_brands_enabled', 'yes' );`.
7. Confirm that the following notice shows up: 

![Markup on 2024-07-25 at 13:38:42](https://github.com/user-attachments/assets/831cfc2c-53da-4ce9-a3c2-45c3a45fa2ae)

8. Confirm that WooCommerce Brands get deactivated. 
9. Confirm that the following notice shows up next to the WooCommerce Brands plugin: 

![Markup on 2024-07-25 at 13:06:45](https://github.com/user-attachments/assets/07305870-aea6-40bb-990f-cc7b8ffce34c)

10. Try to activate it and ensure that this notice shows up: 

![Markup on 2024-07-25 at 16:28:35](https://github.com/user-attachments/assets/ac7eff95-7210-433b-9e5f-88fd854d0b7a)

11. Try to activate WooCommerce Brands via the WP CLI (both with and without the `--skip-plugins` flag) and ensure that it can never be activated. 

12. Update `update_option( 'wc_feature_woocommerce_brands_enabled', 'no' );` in order to disable core's Brands functionality. 
13. Ensure that you can enable the Brands plugin without issues and you see a **Brands** tab under **Products**.
14. Repeat step 1 and ensure that the Brands plugin gets deactivated. 
15. Roll WooCommerce back to 9.3. 
16. Enable WooCommerce Brands and ensure that all the previously created Brands are there. 
17. Update WooCommerce to the latest `trunk`. 
18. Ensure that WooCommerce Brands is still active and all the previously created Brands are there. 
19. Install the [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks/releases/tag/v11.7.0) plugin and ensure that it can be activated without issues.

Part 2: Products > Brands

1. Go to **Products > Brands** and ensure that the Brands you previously created are still there.
2. Create a new brand and add a description and a thumbnail. 
3. Go to **Products > All Products** and create a new product or edit an existing one. 
4. Locate the Brands section on the right hand side (below Product Tags) and assign a brand to the product. 
5. Go back to **Products > Brands** and click to view the Brand that you assigned to that product. 
6. Ensure that the product shows up there. 

Part 3: Shortcodes

1. Create at least 3 Brands under **Products > Brands**.
2. Create a new page under Pages. 
3. Give a title to the page. 
4. Insert this shortcode: `[product_brand_list show_empty_brands="true"]`. 
5. View the page and ensure that all the brands in your store show up alphabetically. 
6. Edit the page. 
8. Insert this shortcode: `[brand_products brand="woo" per_page="12" columns="4"]` and replace `woo` with the slug of one of your brands. 
9. View the page and ensure that the products that belong to this brand show up. 
10. Edit the page. 
11. Insert this shortcode: `[product_brand_thumbnails number="12" show_empty="false"]`.
20. View the page and ensure that all brands thumbnails show up. 
21. Edit the page. 
22. Insert this shortcode: `[product_brand_thumbnails_description number="12" show_empty="false"]`.
23. View the page and ensure that all brands thumbnails along with their descriptions show up. 
24. Edit a product that belongs to a brand. 
25. Insert this shortcode: `[product_brand class="alignright"]` in its short description. 
26. View the product page and ensure that the brand's image shows up. 

Part 4: Widgets

1. Use a theme that supports widgets. 
2. Go to **Appearance > Widgets**.
3. Add the **WooCommerce Brand Description** widget.
4. View a brand's page. 
5. Ensure that the brand's description shows up where you placed the widget. 
6. Add the **WooCommerce Brand Thumbnails** widget.
7. View any page in your store. 
8. Ensure that a list of thumbnails of all brands show up where you placed the widget.
9. Add the **WooCommerce Brand Layered Nav** widget. 
10. View the Shop page. 
12. Ensure that a list of all brands shows up where you placed the widget.

Part 5: Blocks

1. Use a theme that supports blocks. 
2. Go to **Appearance > Editor**.
3. Edit the Single Product template. 
4. Add the Brands block. 
5. View a product that belongs to a brand. 
6. Ensure that the brand's name shows up where you placed your block. 

Part 6: REST API
Go through all these examples: https://woocommerce.com/document/woocommerce-brands/#rest-api-support and ensure that Brands are created, deleted, retrieved as expected.

Part 7: New Product Editor
1. Enable the new Product Editor under WooCommerce > Settings > Features > Advanced. 
2. Edit a product. 
3. Go to the Organization tab. 
4. Ensure that a block where you can select a Brand shows up under the Category selection.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

Introduced Product Brands.

</details>

jimjasson committed:
Move WooCommerce Brands into core

jimjasson committed:
Fix linting errors in brand-thumbnails.php

jimjasson committed:
More lint fixes for brand-thumbnails.php

jimjasson committed:
Fix lint issues in brand-description.php

jimjasson committed:
Fix lint errors for brand-thumbnails-description

jimjasson committed:
Lint errors: Fix taxonomy-product_brand

jimjasson committed:
Lint errors: fix taxonomy-product_brand

jimjasson committed:
Linting: Try adding a space between ignore command and docblock

jimjasson committed:
Another try to remove the lowercase file name error

jimjasson committed:
Another try removing lint error for lowercase files

jimjasson committed:
Lint errors: Fix brands-a-z.php

jimjasson committed:
More lint fixes for brands-a-z.php

jimjasson committed:
More lint fixes for brands-a-z.php

jimjasson committed:
Lint fixes: brand-description.php

jimjasson committed:
Another try fixing lint errors for brand-description.php

jimjasson committed:
Another try fixing lint errors for brand-description.php

jimjasson committed:
More fixes for brand-description.php

jimjasson committed:
Fix lint errors for Packages.php

jimjasson committed:
More fixes for Packages.php

jimjasson committed:
Linting fixes for Brands.php

jimjasson committed:
Added docblocks for WC_Widget_Brand_Thumbnails variables

dakota committed:
Add script to fix coding standards for files changed in branch

dakota committed:
Better messages for fix-branch.sh

jimjasson committed:
Run autofix script for linting

jimjasson committed:
Lint fixes for class-wc-widget-brand-thumbnails.php

jimjasson committed:
More lint fixes for: class-wc-widget-brand-thumbnails.php

jimjasson committed:
More lint fixes for class-wc-widget-brand-thumbnails.php

jimjasson committed:
Lint fixes for class-wc-widget-brand-nav.php

jimjasson committed:
lint fixes: ignore docblocks

jimjasson committed:
Another try to fix missing docblocks

jimjasson committed:
Another try to fix missing docblocks

jimjasson committed:
Another try fixing missing docblocks

jimjasson committed:
Fix lint errors in class-wc-widget-brand-description.php

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

jimjasson committed:
Fix linting errors in REST API and functions classes

jimjasson committed:
Fix linting issues in class-wc-brands.php

jimjasson committed:
More lint fixes for wc-brands.php

jimjasson committed:
More lint fixes for wc-brands.php

jimjasson committed:
Fix lint errors for wc-brands-coupons.php

jimjasson committed:
Fix lint errors for class-wc-brands-block-templates.php

jimjasson committed:
Fix linting errors for class-wc-brands-block-template-utils-duplicated.php

jimjasson committed:
Fix lint errors in class-wc-admin-brands.php

jimjasson committed:
More fixes in class-wc-admin-brands.php

jimjasson committed:
More class-wc-admin-brands.php

jimjasson committed:
More lint fixes for: class-wc-admin-brands.php

jimjasson committed:
More lint fixes for class-wc-admin-brands.php

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Transfer unit test

jimjasson committed:
Transfer e2e test

jimjasson committed:
Added specific versions to templates

jimjasson committed:
Added changelog

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Another try for HTML templates version

jimjasson committed:
Fix lint errors in test files

jimjasson committed:
More lint fixes

jimjasson committed:
Fix lint warnings

jimjasson committed:
Added brands to list of expected REST API fields

jimjasson committed:
More lint warning fixes

jimjasson committed:
More lint warning fixes

jimjasson committed:
Updated unit tests to include brands

jimjasson committed:
Remove whitespace

jimjasson committed:
Added declare( strict_types = 1); to all PHP files

jimjasson committed:
Added declare( strict_types = 1) to test file as well

jimjasson committed:
Fix: There must be exactly one blank line after the file comment

jimjasson committed:
Temporarily remove Brands e2e tests

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Move Brands blockified templates

jimjasson committed:
Remove script to fix lint errors in current branch

jimjasson committed:
Try removing pull-package-deps

jimjasson committed:
Bring back deps

jimjasson committed:
Commit pnpm-lock.yaml

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Add debug statements

jimjasson committed:
More debug statements

jimjasson committed:
Make regular expression more specific

jimjasson committed:
Make matches more specific

jimjasson committed:
Search only for PHP templates

jimjasson committed:
Bring back whitespace

jimjasson committed:
Remove unnecessary change

jimjasson committed:
Update pnpm-lock.yaml

jimjasson committed:
Change the way Brands files are included

jimjasson committed:
Include all files

jimjasson committed:
Prevent Brands assets from being double-enqueued

jimjasson committed:
Move Brands scripts handling into core

jimjasson committed:
Revert changes in the template-changes.ts script

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Use strict in_array

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Add scaffolding for Brands test

jimjasson committed:
Add more scaffolding for Brands tests

jimjasson committed:
Enhance e2e test by adding steps for creating a Brand

jimjasson committed:
Move Brands test to Playwright folder

jimjasson committed:
Added manifest

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Fix lint errors in tests

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

jimjasson committed:
Move Brands coupons test into core's coupons tests

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Fix linting error in tests

ObliviousHarmony commented on plugins/woocommerce/src/Packages.php:
```diff
@@ -47,20 +47,41 @@ private function __construct() {}
+	 * @var array Key is the package name/directory, value is the main plugin class which handles init.
+	 */
+	protected static $merged_plugins = array(
+		'woocommerce-brands' => '\\Automattic\\WooCommerce\\Internal\\Brands',
```
How would you feel about changing the name of `$merged_packages` instead? The point of the feature in the first place was because the _plugin_ was merged after all!

jimjasson replied:
Hey @ObliviousHarmony! 

The main reason I went with a different array/name here is that the `initialize_packages` [function](https://github.com/woocommerce/woocommerce/blob/c11a14629cacc7b32a2882520014c6eda2a12398/plugins/woocommerce/src/Packages.php#L205-L235) seems to be doing a lot more than [what I need](https://github.com/woocommerce/woocommerce/blob/c11a14629cacc7b32a2882520014c6eda2a12398/plugins/woocommerce/src/Packages.php#L242-L246) for Brands. Do you think it is OK to let `initialize_packages` to run for my own, "dummy" package or do you have any other suggestions on how to run only part of `initialize_packages` for Brands?

jimjasson replied:
I also need plugins, like Brands, in a separate array to run functions like these: 
- https://github.com/woocommerce/woocommerce/pull/50165/files#diff-c3a93b21b4c564d9c866729f93fcaf38838c2a67ab30e5cf8a0bd4db7e269fdcR147
- https://github.com/woocommerce/woocommerce/pull/50165/files#diff-c3a93b21b4c564d9c866729f93fcaf38838c2a67ab30e5cf8a0bd4db7e269fdcR169
- https://github.com/woocommerce/woocommerce/pull/50165/files#diff-c3a93b21b4c564d9c866729f93fcaf38838c2a67ab30e5cf8a0bd4db7e269fdcR188

ObliviousHarmony replied:
One thing to keep in mind is that we're currently using this to do the same thing with WooCommerce Blocks and WooCommerce Admin's old plugins. It should disable the feature plugin and prevent it from being activated. The UX for this behavior in this PR is awesome and I think these old merged plugins have the same benefits.

The only thing that `$merged_packages` seems to do differently is that it calls `init()` on the plugin's main class file. We could probably have it do the same and keep any initialization in the associated class file. You'd just have to aim at it!

With that in mind, it _does_ seem like it might make sense to replace the "merged_packages" naming with "merged_plugins" and avoid the duplicate code? As noted, since the UX of the disabling, prevention, and plugin list is better in this PR, we could remove the existing package disabling and prevention behavior with this.

What do you think?

dakota replied:
Hey @ObliviousHarmony, @jimjasson is AFK this week, so I'm hopping on here.

As a fresh pair of eyes, I think it makes a lot of sense to change `$merged_packages` to `$merged_plugins` and use the UX introduced in this PR.  

I see that `$merged_plugins` is also calling the `init()` method on the plugin's main class, so it does seem like duplicate code.

I'll take a stab at cleaning it up and de-dupping things.

dakota replied:
Changes done; please review them :) See https://github.com/woocommerce/woocommerce/pull/50165/files/b320687c68023aa82b045c5de52adcedb93d2160..ff07d4c566951fbde0f7265ec099a58c1bd6a4da

ObliviousHarmony replied:
Thanks @Dakota, that looks great!

jimjasson committed:
Move all Brands initialization within the /Internal/Brands class

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

# Conflicts:
#	docs/docs-manifest.json

budzanowski commented:
I am quite opinionated on how we should structure such add-ons to core so please don't treat my suggestion as criticism but just as an alternative approach.

I have formed this opinion based on the following article:

[Kernighan and Pike were right: Do one thing, and do it well](https://medium.com/source-and-buggy/do-one-thing-and-do-it-well-886b11a5d21)

and observing the core development.

IMO we should treat the functionality still as a plugin. This means that the whole codebase should live under a separate folder and interact with core via APIs ( hooks, filters, core function calls  ) - just as it is doing right now.

Why?

- tracking Brands functionality in core becomes simple
- WC core remains lean ( it should be even leaner IMO but that is a different discussion )
- lean core provides better overall stability
- disabling the feature is easy ( we need this as an option ) - what if someone wants to use a different Brands type plugin?

> However, in the future, it is worth exploring the option of moving away from a hook-based system and instead move the Brands code inside existing files in core.

With the more in core initiative this using approach we would quickly corner our selves into oversized core that does all - maintenance and quality would be much harder.

This is the first element of the more in core initiative - it will set a standard of how we will approach all of the remaining plugins.

As I have said this is an alternative approach. The work you did in this PR looks splendid if this is how we want to do it :)

dakota committed:
Rename `$merged_packages` to `$merged_plugins`

dakota committed:
Add force disable method back

vedanshujain commented:
Closing and reopening the PR to trigger CI.

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Move Brands logic outside core files

jimjasson committed:
Rename admin styles

jimjasson committed:
Remove brands logic from core's admin class

jimjasson committed:
Roll back all changes in admin assets class

jimjasson committed:
Fix linting errors

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Move REST API logic to Brands main class

jimjasson committed:
Introduce an option to control whether the Brands package is enabled.
Prevent autoloader from loading classes already loaded by individual Packages.
Fix an issue with Brands admin styles.

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

# Conflicts:
#	pnpm-lock.yaml

jimjasson committed:
Bring back pnpm-lock

jimjasson committed:
Add comment

jimjasson committed:
Split long line into two

vedanshujain commented on plugins/woocommerce/src/Packages.php:
```diff
@@ -74,6 +94,61 @@ public static function package_exists( $package ) {
+		foreach( self::$merged_packages as $merged_package_name => $package_class ) {
+
+			// For gradual rollouts, ensure that a package is enabled for user's remote variant number.
+			$experimental_package_enabled = method_exists( $package_class, 'is_enabled' ) ?
```
optional nitpick, but maybe it makes more sense to have the default as `false`. we do want packages to have `is_enabled` method.

jimjasson replied:
I was thinking that perhaps for feature packages we won't want to do a gradual rollout, but let's switch to a `false` default value for now and re-visit if this becomes an issue in the future.

vedanshujain commented on plugins/woocommerce/src/Internal/Brands.php:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	public static function is_enabled() {
+		$assignment = get_option( 'woocommerce_remote_variant_assignment', 0 );
```
let's set this to false for the first release, and then we will enable it in a subsequent one, wdyt? also, we should probably pick some random start number which is not `0`, otherwise for some shops, all the experimental features will be enabled at once.

jimjasson replied:
Good idea to have the default value to `false` and prevent stores that don't have this option from loading this package. 

Regarding picking a random number: 
- the `is_enabled` function is isolated and doesn't update the `woocommerce_remote_variant_assignment` and;
- the main WooCommerce file is responsible for picking a random number for `woocommerce_remote_variant_assignment` and then updating the option: [here](https://github.com/woocommerce/woocommerce/blob/057e118a3443721d947aa92c0487aac1238d288a/plugins/woocommerce/includes/class-woocommerce.php#L309-L310) and [here](https://github.com/woocommerce/woocommerce/blob/057e118a3443721d947aa92c0487aac1238d288a/plugins/woocommerce/includes/class-woocommerce.php#L370).

That said, I don't think that individual packages should update this option, as this controls many experimental features and not only Brands.

vedanshujain reviewed (commented):
Adding some comments for the merge infra.

jimjasson committed:
Review default values for remote variant assignment

jimjasson committed:
Rename global functions and add polyfills for deprecated functions

jimjasson committed:
Bump versions

jimjasson committed:
Fix some lint errors

jimjasson committed:
More lint fixes

jimjasson committed:
Set woocommerce_remote_variant_assignment for Brands to be enabled for unit tests

jimjasson committed:
Replace reserved word class with class_name

jimjasson committed:
Another try to include Brands files in tests

jimjasson committed:
Remove Brands from REST API tests

jimjasson committed:
Skip Brands tests while Brands is behind a feature flag

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Lint fixes

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

jimjasson committed:
Remove empty line

jimjasson committed:
Added feature flag.

jimjasson committed:
Fix widgets form

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Fix lint errors for brand description widget

jimjasson committed:
Fix lint errors for brand description widget

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

# Conflicts:
#	plugins/woocommerce/includes/widgets/class-wc-widget-brand-description.php

jimjasson committed:
Fix lint errors

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

lanej0 commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-product-brand.spec.js:
```diff
@@ -0,0 +1,181 @@
+
+test.use( { storageState: process.env.ADMINSTATE } );
+
+test.skip( 'Merchant can add brands', async ( { page } ) => {
```
Is it intentional that this test is skipped?

jimjasson replied:
@lanej0 In v9.4, Brands will be behind a feature flag that will be disabled by default. Therefore, Brands logic is not loaded while E2E tests run, which is why we decided to skip them for now. When the feature flag is removed, we will run these tests again. 

A similar approach was followed in other cases, where features were behind a feature flag + there were tests for these features: 

1. [Disable Launch Your Store feature flag in 9.0](https://github.com/woocommerce/woocommerce/pull/48231)
2. [Enable Launch Your Store feature flag in 9.1](https://github.com/woocommerce/woocommerce/pull/48244)

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson commented:
> When running the e2e tests locally, I had 3 failures in the shopper flows:

Thank you for raising this! 

I had a look at the tests you mentioned and they don't seem related to Brands -- as you said, an updated selector may be triggering these errors. Given that this is not related to this PR, I propose that this gets addressed in a separate issue/PR.

vedanshujain commented on plugins/woocommerce/includes/blocks/class-wc-brands-block-template-utils-duplicated.php:
```diff
@@ -0,0 +1,367 @@
+ * BlockTemplateUtils class used for serving block templates from Woo Blocks.
+ * IMPORTANT: These methods have been duplicated from Gutenberg/lib/full-site-editing/block-templates.php as those functions are not for public usage.
+ *
+ * @version x.x.x
```
let's change the version strings, also let's add a note somewhere that these are internal use for Brands code and not meant to be public API.

vedanshujain reviewed (commented):
looks good to me, note that the testing instructions needs to be changed to remove the variant assignment options. some more test cases that we should keep in mind as part of testing after merge:
- [x] With other packages, namely WC Blocks and WC Admin standalone plugin. Perhaps just activation flows with WC Blocks testing since the standalone admin plugin was archived quite a while ago. 
- [x] Upgrade path from WooCommerce 9.3 with brands to this build with brands.
- [x] WooCommerce 9.4 with brands rollback to WooCommerce 9.3.

jimjasson committed:
Bump version

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Updated tooltips for Brands coupon restrictions to match core's

jimjasson committed:
Fix lint errors

jimjasson committed:
More lint fixes

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

jimjasson committed:
Add REST API v3 for Brands

jimjasson committed:
Merge branch 'merge-brands-in-core' of https://github.com/woocommerce/woocommerce into merge-brands-in-core

jimjasson committed:
Merge branch 'trunk' into merge-brands-in-core

xristos3490 reviewed (approved):
Great work, @jimjasson!

I've tested the following:
- Confirmed that the initial state does not include the feature.
- Confirmed that enabling the feature by adding the `wc_feature_woocommerce_brands_enabled` option works as expected.
- Confirmed that enabling the feature while having the Brands extension installed and activated successfully deactivates the extension and adds a notice there.
- Confirmed that behind the scenes, the variant option is still applicable, giving us a good transition in the future.

I've also reviewed the code changes, mainly on the Packages controller and the autoloader, and they seem to be working as expected.

Just a heads-up:
- I considered the Brands functionality out of the scope of this PR. While I tested basic flows of the extension, I didn't go into bug hunting mode on the functionality, and I didn't expect this PR to bring improvements to the extension itself.
- The file structure used here to merge the extension follows a path that has been discussed by numerous people, and we all reached a consensus that these file locations are well integrated in the `includes` folder.

With all that being said, I'm giving my approval for the PR! Once again, great job handling this, Jason!

anthonyattard commented:
This is causing integration issues with custom plugins that implement this feature along with custom product schema while the woo core does not implement product schema choosing still to have it paywalled.
